typing/jkind.ml

```diff
@@ -1457,94 +1466,105 @@ module Jkind_desc = struct
-                  (With_bounds.to_list b_jkind.jkind.with_bounds)
-              in
-              (* Use *original* ctl here, so we don't fall over on
-                  a record with 20 lists with different payloads. *)
-              let bounds, bs' = loop ctl_after_unpacking_b bounds_so_far bs in
-              bounds, With_bounds.join_bounds_maybe_empty nested_with_bounds bs'
-            else
-              (* Don't know the best kind for this type, so skip it, leaving it in the
-                 (normalized) with_bounds. *)
-              let bounds_so_far =
-                if require_best
-                then bounds_so_far
-                else join_respecting_omit Bounds.max
+              let bounds_so_far, nested_with_bounds, fuel_result1 =
+                loop new_ctl bounds_so_far (With_bounds.to_list b_with_bounds)
               in
-              let bounds_so_far, bs' =
-                loop ctl_after_unpacking_b bounds_so_far bs
+              (* CR: which ctl to use?
+                 (* Use *original* ctl here, so we don't fall over on
+                     a record with 20 lists with different payloads. *)
+              *)
+              let bounds, bs', fuel_result2 = loop new_ctl bounds_so_far bs in
+              ( bounds,
+                With_bounds.join nested_with_bounds bs',
+                Fuel_status.both fuel_result1 fuel_result2 )
+            | Not_best, Require_best ->
+              let bounds_so_far, bs', fuel_result =
+                loop new_ctl bounds_so_far bs
               in
-              bounds_so_far, With_bounds.add_bound_maybe_empty ty ti bs'
-          | None ->
-            (* kind of b is not principally known, so we treat it as having the max
-               bound (only along the axes we care about for this type!) *)
-            let bounds_so_far = join_respecting_omit Bounds.max in
-            let bounds_so_far, bs' =
-              loop ctl_after_unpacking_b bounds_so_far bs
-            in
-            bounds_so_far, With_bounds.add_bound_maybe_empty ty ti bs'))
+              bounds_so_far, With_bounds.add ty ti bs', fuel_result
+          in
+          match Loop_control.check ctl ty with
+          | Stop ctl_after_stop ->
+            (* out of fuel, so assume [ty] has the worst possible bounds. *)
+            found_jkind_for_ty ctl_after_stop Bounds.max No_with_bounds Not_best
+          | Skip -> loop ctl bounds_so_far bs (* skip [b] *)
+          | Continue ctl_after_unpacking_b -> (
+            match jkind_of_type ty with
+            | Some b_jkind ->
+              found_jkind_for_ty ctl_after_unpacking_b b_jkind.jkind.mod_bounds
+                b_jkind.jkind.with_bounds b_jkind.quality
+            | None ->
+              (* kind of b is not principally known, so we treat it as having the max
+                 bound (only along the axes we care about for this type!) *)
+              found_jkind_for_ty ctl_after_unpacking_b Bounds.max No_with_bounds
```

Comment:
Might be hard to do, but I don't think there are any tests for this case.

Reply:
I've done this in @glittershark's normalization pr in commit e8190bd6f921f1cc5ac2b9d7cda6738afac61a7d : https://github.com/ocaml-flambda/flambda-backend/pull/3509/commits/e8190bd6f921f1cc5ac2b9d7cda6738afac61a7d